Draggable widget: allow to hide the original widget when dragging.

Deltaspace0 opened the pull request:
When draggable widget is dragged, the item is rendered twice: in the original place and under the mouse pointer.
I added `draggableHideOrigin` config option for draggable widget which hides the original widget and renders it only under the pointer.

Without `draggableHideOrigin` (by default):

![default](https://user-images.githubusercontent.com/26405676/235747859-4527765f-851c-4fa0-af2f-8ff18d9b1031.gif)

With `draggableHideOrigin`:

![hide-origin](https://user-images.githubusercontent.com/26405676/235747847-74060fb4-8b4f-4e39-9d9f-91942558bcb9.gif)

Deltaspace0 commented:
I have come up with alternative names for this configuration: `hideChild`, `invisibleChild`. Also I thought about making it the default behavior and instead add `showChild` configuration.

fjvallarino commented:
@Deltaspace0 sorry for the delay. I wanted to think a bit more about finding a way to avoid rendering the source widget when it's hidden. 

One option is doing something similar to that `tooltip` does in its [render](https://github.com/fjvallarino/monomer/blob/main/src/Monomer/Widgets/Containers/Tooltip.hs#L197) function. In this case, instead of reusing the infrastructure provided by `Container`, it provides its own complete version of `render`. While `Container`'s `render` function does a lot of other things, sometimes you can just get by with a basic implementation. For example, in the scenario you want to implement, you could do something like:

```haskell
  render wenv node renderer = do
    when (not hideOrigin) $
      forM_ children $ \child ->
        widgetRender (child ^. L.widget) wenv child renderer

    when dragged $ do
      createOverlay renderer $ do
        renderAction config wenv node renderer
    where
      dragged = isNodeDragged wenv node
      renderAction = fromMaybe defaultRender (_dgcCustomRender config)
      hideOrigin = fromMaybe False (_dgcHideOrigin config)
```

It's important to note that `tooltip` does not implement the `render` function accepted by `Container`; it instead overwrites the one in `Widget`:

```haskell
  baseWidget = createContainer state def {
    ...
  }
  widget = baseWidget {
    widgetRender = render
  }
```

In case you prefer your current solution, I'm happy with it too.

---

I don't have strong feelings about keeping the default behavior of rendering the source widget, so feel free to change it. May I suggest naming the config option `draggableRenderSource` instead?

Deltaspace0 commented:
It didn't cross my mind that it is possible to just not render the child widget by overwriting the `widgetRender`. I changed the solution, it is definitely more proper approach. Also, I replaced `draggableHideOrigin` with `draggableRenderSource` and now the source widget does is not rendered when dragging by default.

fjvallarino commented:
@Deltaspace0 looks great, thanks! 

The documentation needs some work to make this type of thing discoverable. Right now, unless you read the source code of the existing widgets and happen to find one that does what you need, it's really hard to know if it is possible.

Beyond this, I will also add some more extension points in `Container` to make this and related things simpler.
